cdisutils/parsers.py

```diff
@@ -0,0 +1,80 @@
+import re
+
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
+
+
+class S3URLParser:
+    """Adapter for urlparse to facilitate working with S3 urls"""
+    def __init__(self, s3_url):
+        self._s3_url = s3_url
+        self.scheme = None
+        self.netloc = None
+        self.bucket = None
+        self.key = None
+        self._parse()
+
+    def __repr__(self):
+        return "S3URLParser(scheme='{}', netloc='{}', bucket='{}', key='{}')".format(
+            self.scheme,
+            self.netloc,
+            self.bucket,
+            self.key,
+        )
+
+    def _parse(self):
+        """Parsing options for new and old style s3 urls.
+
+        All new buckets created on aws will use the new style of url.
+
+        Old: s3://url/bucket_name/key/name/at/the/end
+        New: https://bucket.url/key/name/at/the/end
+        """
+        parse_object = urlparse(self._s3_url)
+
+        self.scheme = parse_object.scheme
+        self.netloc = parse_object.netloc
+
+        # (scheme, bucket, netloc, path)
+        aws_regex = re.compile(r"^(s3|https?):\/\/([^\.\s]+)\.?(s3.*\.amazonaws\.com)\/(.+)$")
```

Comment:
I agree that a regex with capture groups is the most natural way to handle this case. I have a few very minor implementation comments:

1. The first dot should not be optional, or else this can happen:

    ```
    >>> parsers.S3URLParser("s3://yes3.amazonaws.com/part1/part2")
    S3URLParser(scheme='s3', netloc='s3.amazonaws.com', bucket='ye', key='part1/part2')
    ```

2. `match` only matches at the beginning of the string, and the last capture group matches everything after the slash, so I don't think the anchors are needed.
3. `re.compile` doesn't really do anything here because it'll be called every time this function executes. So, with or without that explicit call, this function relies on the regex compiler to cache things. Doesn't hurt, doesn't help.

Reply:
Good catch. I mentioned it in another comment but I was trying to solve too much with the regex.

> match only matches at the beginning of the string, and the last capture group matches everything after the slash, so I don't think the anchors are needed.

The beginning anchor could be useful still. It's my personal preference to include anchors on regular expressions, regardless of how they begin or end, but If you feel strongly about it I will remove them.

Reply:
I do not feel strongly and am fine with the anchors being there.